### DISCUSSION TITLE
passing variable contents across the experiment

### DISCUSSION THREAD

**Discussion by EceKayaMPI** [jspsych/jsPsych#925] [Q&A] [answered]
Hi, 
I have been trying to setup my experiment such that each participant gets a different pre-defined stimuli list. 
I managed to get the participant_ID, given by jatos and/or prolific into the code. However, I could not pass the variable contents inside another plugin, the variable environment or anything. I can only import the participant_ID inside 'on_start' functions, and can even see it on the console if I push the contents to some variable (e.g., stims.push(audio_stims) and then see the contents in console.log) BUT I can not access that 'stim' variable somewhere else.
Interestingly, I can see the variable contents in 0: if I do console.log(stims), but it gives an empty array if I try to get 
var stims_to_use = stims[0]

I have tried accessing the batch/jatos/prolific etc. info in the general script. It only works in on_start functions. 
Can anyone help me? 

Here is the code: https://github.com/EceKayaMPI/study_assets_root/blob/master/wobbles/CA_batchexample.html

Thanks in advance, 
Ece

**Comment by EceKayaMPI** [comment, marked as answer]
TL;DR: is there a way to use user input or batch session data (retrieved from URL) in the experiment? 
To my experience, the variables are initialised before all other functions etc.., and cannot be changed afterwards.

**Reply by jodeleeuw** [reply]
I'm not sure I fully understand the goal here, but hopefully this is a helpful comment:

If you are retrieving URL data, then that data is available during the setup phase for the experiment. So you can do something like:

```js
var condition = jsPsych.data.getURLVariable('condition');
if(condition == 0){
	var stimuli = stimuli_list[0];
	// or whatever else you need to do...
}
if(condition == 1){
	var stimuli = stimuli_list[1];
}

// later...

var procedure = {
	timeline: [ ... ],
	timeline_variables: stimuli
}
```

**Reply by EceKayaMPI** [reply]
Hi,
I have tried getting URL data as in the example, but the console logs this as undefined:

var subj_ID = jsPsych.data.getURLVariable('prolific_id=')
console.log(subj_ID)

The link is:
http://localhost:9000/publix/5/start?batchId=15&generalMultiple&prolific_id=1234567

**Reply by jodeleeuw** [reply]
Leave out the `=` part and it should work:

```js
var subj_ID = jsPsych.data.getURLVariable('prolific_id')
console.log(subj_ID)
```

**Reply by EceKayaMPI** [reply]
Unfortunately it doesn't work. I have tried all parts of that string to get the URL variable.
The simple code where I try to log the URL variable is here: https://github.com/EceKayaMPI/study_assets_root/blob/SIMPL/wobbles/CA_batchexample.html

My whole study depends on this information I have been trying to retrieve for weeks now :(

**Reply by EceKayaMPI** [reply]
I think the problem is here:
When the link above is clicked, it changes something like this:
http://localhost:9000/publix/5/10/start?srid=404
And I think this is where jsPsych executes. Because if I request all URL variables with jsPsych.data.urlVariables(), it retrieves srid=404. 
Since I use JATOS, I have to add the jatos function above the jsPsych, where I can also retrieve the prolific ID as well. However, for the experiment to work, I need to be able to pass it into jsPsych variable environment. Is there a way to do it?

**Reply by jodeleeuw** [reply]
Hmm. If that example isn't working then I don't think the jsPsych code is the problem here. A couple possibilities:

- Seems unlikely, but maybe JATOS interferes with this code in some way. @becky-gilbert might know if there are issues like this? Or @kristian-lange? 
- Make sure to test with a valid query string. Seems like you are using a valid string, but adding this for completeness!

**Reply by jodeleeuw** [reply]
Ah OK. So that sounds like it is a JATOS issue. I'm not super familiar with how to work with JATOS. Hopefully Becky or Kristian can offer some advice here!

**Reply by kristian-lange** [reply]
Hi everyone,
I'm a bit blurry about the question but I'll try my best to help.
When JATOS starts a study, it redirects to a new URL and all URL query parameters are available via `jatos.urlQueryParameters`. So in your case the URL query parameter `prolific_id` is available in `jatos.urlQueryParameters.prolific_id`. I do not know the internals of jsPsych but my guess is that `jsPsych.data.getURLVariable('prolific_id')` does not work because the URL query parameters are not part of the URL anymore (JATOS stripped them off and put them into jatos.urlQueryParameters). Does this help already?
Best, 
Kristian

**Reply by EceKayaMPI** [reply]
Thanks a lot for your help anyway. I am working on a workaround:
In the first trial: 

on_start: function () {
              var current_particip = getPROid(); // get ID here. It works when it is a start function.
              jsPsych.data.write(current_particip) // add to data 
....

And then retrieve it from jsPsych.data.get() later, but I can't see it getting added there...

**Reply by EceKayaMPI** [reply]
> Hi everyone,
> I'm a bit blurry about the question but I'll try my best to help.
> When JATOS starts a study, it redirects to a new URL and all URL query parameters are available via `jatos.urlQueryParameters`. So in your case the URL query parameter `prolific_id` is available in `jatos.urlQueryParameters.prolific_id`. I do not know the internals of jsPsych but my guess is that `jsPsych.data.getURLVariable('prolific_id')` does not work because the URL query parameters are not part of the URL anymore (JATOS stripped them off and put them into jatos.urlQueryParameters). Does this help already?
> Best,
> Kristian

Hi Kristian, thanks for your help. However, I suggested these as the source of the problem in my previous comment already. 
Jatos functions successfully retrieve the needed information. The **problem** is that jsPsych variables are initiated before I can retrieve and use Jatos information, and thus, I can only use these variables inside on_start functions for trials, but can't pass the information to the variable environment (where the timeline variable etc.. are initiated). 

The **solution** I am looking for is a way to pass a variable that is generated/retrieved before execution of jsPsych.init inside jsPsych variable environment. 

PS: I have also tried _creating an empty variable and use .push_ in every possible point of the code.

**Reply by jodeleeuw** [reply]
Can you move your jsPsych experiment generation code inside the jatos.onLoad function?

**Reply by EceKayaMPI** [reply]
It is inside a function that executes conditionally inside the jatos.onload function. 
https://github.com/EceKayaMPI/study_assets_root/blob/SIMPL/wobbles/CA_batchexample.html

**Reply by jodeleeuw** [reply]
It looks like `jsPsych.init()` is called inside that function, but that the timeline `timeVar` is built outside that function. Can you construct the timeline inside `jatos.onLoad` too?

**Reply by EceKayaMPI** [reply]
> It looks like `jsPsych.init()` is called inside that function, but that the timeline `timeVar` is built outside that function. Can you construct the timeline inside `jatos.onLoad` too?

This works!! (I also had to bring init_experiment up)
Thank you so much :) You are a life-saver!
The final, working code is here: 
https://github.com/EceKayaMPI/study_assets_root/blob/SIMPL/wobbles/CA_batchexample.html

Cheers,
Ece
